# TDL 8.1 Thinking with Deep Learning: Week 8 Part 1
# Sampling and Reliability 

__Instructor:__ James Evans

__Teaching Assistants & Content Creators/Organisers:__ Bhargav Srinivasa Desikan (Notebook Author), James Evans (Notebook Editor), Likun Cao (Notebook Reviewer)

In this notebook we will explore some crucial aspects of dealing with data - sampling, annotations, and reliability. This will lay the groundwork for when we begin to work with Active Learning.


In [ ]:
# empty cell

# Sampling 

[Sampling](https://en.wikipedia.org/wiki/Sampling_(statistics)) is a popular method which we've likely come across in our previous data science or social science explorations. In this section we will be exploring some of the more popular techniques that are used in industry and research. There is no one standard package for sampling with python, and we will be using the rich PyData ecosystem in different ways to get what we want done.

Here are some links you may wish to explore:

- Krippendorff, Klaus. 2004. Content Analysis: An Introduction to its Methodology. Thousand Oaks, CA: Sage: [“Sampling”](https://canvas.uchicago.edu/courses/33672/files/4767016/download?wrap=1) 
- [Data Scientist's guide to 8 sampling techniques](https://www.analyticsvidhya.com/blog/2019/09/data-scientists-guide-8-types-of-sampling-techniques/
)
- [KDnuggets - 5 sampling algorithms](https://www.kdnuggets.com/2019/09/5-sampling-algorithms.html)

Sampling procedures are often divided into probabilistic and non-probabilistic methods, and we will start with the same division before jumping into more machine and deep learning relevant methods.




## Probabilistic Sampling

A lot of times we hear about sampling in the context of sampling from a probability distribution. Such methods can often, if not always, be extended to sampling from real world data. For the examples in this section, we will use both a real world dataset, as well as sample from random distributions, both of which are useful tools. The power of these methods in the context of machine learning and deep learning (and its human aspects!) will become more clear as we go through them.  

### Dataset

One of the datasets we will be using is the "Mental Health in Tech Survey" data, which is an open source survey data about mental health conditions of workers in the tech industries. We also worked with this data for the week 5 hint. You can find the data at Kaggle:

https://www.kaggle.com/osmi/mental-health-in-tech-survey

[google drive link of cleaned data](https://drive.google.com/file/d/1-0r2C4z9-vAedJ4uum-TfI4Zy4gKDGaQ/view?usp=sharing)

We provided a cleaned version of this data. The predictors contain 1 continuous variable (age), 3 dummies (Do you work remotely? Is your employer primarily a tech company? Does your employer provide any mental health benefits?) and 2 categorical variables (gender-male/female/other; can you discuss your mental health issue with supervisors-yes/sometimes/no).

The outcome is an answer to the question: If you have a mental health condition, do you feel it interferes with your work? The DV is measured with a 5-categorical variable: NA (no mental health condition), never, rarely, sometimes, often. 

Don't worry too much about the data - it's just useful for us to have a dataset to refer to when we are sampling. **An important note**: very often we are sampling from a very large dataset or population - in this case, our dataset is small enough that we can use the whole dataset, and the sampling is purely illustratory.



In [ ]:
import pandas as pd

In [ ]:
df=pd.read_csv('/content/mental health.csv')

In [ ]:
df

age  remote  benefits  tech  gender    supervisor  interfere
0       8       1         1     1   other           yes          4
1      21       0         0     1   other  some of them          3
2      32       0         0     1   other            no          4
3      28       0         0     1   other  some of them          2
4      27       1         1     1   other           yes          4
...   ...     ...       ...   ...     ...           ...        ...
1254   32       1         1     1   other           yes          4
1255   26       1         0     1   other  some of them          3
1256   30       0         1     1   other           yes          2
1257   18       1         1     1   other           yes          0
1258   34       0         1     0  female  some of them          2

[1259 rows x 7 columns]

### Random



In [ ]:
sample_df = df.sample(100)

In [ ]:
sample_df

age  remote  benefits  tech  gender    supervisor  interfere
896    25       1         0     0    male            no          4
79     28       0         1     1  female           yes          0
1125   35       0         0     1    male           yes          3
214    23       1         1     1    male  some of them          0
644    37       0         1     0    male            no          0
...   ...     ...       ...   ...     ...           ...        ...
1124   25       0         0     1    male            no          0
737    29       0         1     1    male           yes          4
122    26       0         1     1  female           yes          3
236    40       0         1     1    male  some of them          1
915    56       0         1     1   other  some of them          2

[100 rows x 7 columns]

[SciPy](https://docs.scipy.org/doc/numpy-1.10.1/reference/routines.random.html) and [numpy](https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html) also offer many popular sampling techniques, usually to be applied on a list or array, or sampled from a distribution.

In [ ]:
import numpy as np
import scipy as sp

In [ ]:
random_vals = np.random.rand(500)

In [ ]:
random_vals

array([0.43699714, 0.47060992, 0.24905942, 0.22239904, 0.44296175,
       0.22545222, 0.22134929, 0.96157734, 0.62211869, 0.44807806,
       0.50607784, 0.73835084, 0.9380274 , 0.45296616, 0.29843266,
       0.13536347, 0.25888749, 0.05993663, 0.60588416, 0.41785586,
       0.06071734, 0.15378653, 0.88856166, 0.37703047, 0.02557786,
       0.33333586, 0.41360987, 0.09058795, 0.98022939, 0.2889466 ,
       0.05141591, 0.65477716, 0.26518663, 0.43644931, 0.87321368,
       0.94961348, 0.19790819, 0.91365039, 0.46810631, 0.45035853,
       0.42123755, 0.72920887, 0.55388844, 0.7647531 , 0.51071366,
       0.64269204, 0.42339985, 0.73069075, 0.50256078, 0.52905678,
       0.48488404, 0.72724541, 0.40514898, 0.77297362, 0.23124549,
       0.86063728, 0.76898662, 0.4613524 , 0.04014646, 0.1411483 ,
       0.13068955, 0.32412105, 0.63241511, 0.16702814, 0.30841922,
       0.43568952, 0.8862674 , 0.20357862, 0.85145767, 0.76444359,
       0.92301258, 0.69283349, 0.2460816 , 0.25595245, 0.51492

In [ ]:
np.random.choice(random_vals, 10)

array([0.73069075, 0.34131513, 0.80750426, 0.77948091, 0.9920094 ,
       0.95253342, 0.81190901, 0.11929868, 0.75053042, 0.44680208])

In this case, we first generated an array of size 500 by randomly sampling between 0 and 1, and then sampled 10 values from this list. We can similarly sample values from any list by using the scipy and numpy random module.

### Stratified

Stratified random sampling is a method of sampling that involves the division of a population into smaller sub-groups known as strata. In stratified random sampling, or stratification, the strata are formed based on members' shared attributes or characteristics such as, in our example, age or interference. [Here](https://www.investopedia.com/terms/stratified_random_sampling.asp) is a source for reading more about it.

In [ ]:
from sklearn.model_selection import train_test_split


In [ ]:
stratified_sample, _ = train_test_split(df, train_size=0.10, stratify=df[['remote']])

In [ ]:
stratified_sample

age  remote  benefits  tech  gender    supervisor  interfere
1176   38       0         0     1    male           yes          0
86     30       1         0     1  female  some of them          0
470    33       0         1     1    male  some of them          3
866    58       0         1     1    male           yes          2
71     26       1         0     1  female           yes          3
...   ...     ...       ...   ...     ...           ...        ...
221    27       0         0     1    male            no          2
35     31       0         1     1  female           yes          3
206    34       0         1     1    male            no          3
568    30       1         1     1    male            no          0
83     28       0         1     1  female  some of them          3

[125 rows x 7 columns]

In this case we get stratified results for remote work, and our remote work attribute is represented as per the original ratio.

### Varying probability sampling (weighted by variables of interest)

We can also sample by weighting certain attributes so that our sample represents it accordingly. With Pandas we can do this using a DataFrame column as weights. Rows with larger value in the  column are more likely to be sampled.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html


In [ ]:
weight_sample = df.sample(n=125, weights='age')

In [ ]:
weight_sample

age  remote  benefits  tech  gender    supervisor  interfere
578    26       0         0     1    male  some of them          3
502    32       0         0     1    male           yes          3
55     38       0         1     1  female           yes          3
611    23       0         0     1    male            no          3
116    28       0         1     0  female            no          2
...   ...     ...       ...   ...     ...           ...        ...
521    28       0         0     1    male           yes          1
1018   29       0         0     1    male           yes          0
49     31       0         0     1  female  some of them          3
956    34       0         1     0  female  some of them          3
468    31       1         1     1    male           yes          0

[125 rows x 7 columns]

In [ ]:
weight_sample['age'].mean()

32.568

In [ ]:
df['age'].mean()

32.01906274821287

Because we weighted by age, older ages will be prioritised, which leads to that small bump in the mean age of the weighted sample.

### Cluster Sampling (lists of large groups of units)

Cluster sampling is a type of probability sampling in which every and each element of the population is selected equally, we use the subsets of the population as the sampling part rather than the individual elements for sampling.

The population is divided into subsets or subgroups that are considered as clusters, and from the numbers of clusters, we select the individual cluster for the next step to be performed.

You can read more about cluster sampling [here](https://www.geeksforgeeks.org/cluster-sampling-in-pandas/). 

There are a couple of ways to do cluster sampling - one way is to somehow cluster the dataset, and then choose that whole cluster as your sample. In this case, we will cluster based on age and then choose samples from one of these clusters.



In [ ]:
from sklearn.cluster import KMeans

In [ ]:
age_cluster = KMeans(n_clusters=12)

In [ ]:
age_cluster.fit(np.array(df['age']).reshape(-1, 1))

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=12, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=None, tol=0.0001, verbose=0)

In [ ]:
df['cluster'] = age_cluster.labels_

In [ ]:
df

age  remote  benefits  tech  gender    supervisor  interfere  cluster
0       8       1         1     1   other           yes          4       11
1      21       0         0     1   other  some of them          3        4
2      32       0         0     1   other            no          4        3
3      28       0         0     1   other  some of them          2        8
4      27       1         1     1   other           yes          4        1
...   ...     ...       ...   ...     ...           ...        ...      ...
1254   32       1         1     1   other           yes          4        3
1255   26       1         0     1   other  some of them          3        1
1256   30       0         1     1   other           yes          2        8
1257   18       1         1     1   other           yes          0        5
1258   34       0         1     0  female  some of them          2        7

[1259 rows x 8 columns]

In [ ]:
df[df['cluster'] == 3]

age  remote  benefits  tech  gender    supervisor  interfere  cluster
2      32       0         0     1   other            no          4        3
5      33       0         0     1   other            no          3        3
8      31       0         0     1   other            no          3        3
23     33       0         0     1  female           yes          3        3
27     32       0         0     1  female  some of them          3        3
...   ...     ...       ...   ...     ...           ...        ...      ...
1229   31       0         1     1    male  some of them          3        3
1238   32       0         0     1    male            no          0        3
1247   32       0         1     1    male            no          3        3
1250   32       1         0     1   other            no          3        3
1254   32       1         1     1   other           yes          4        3

[224 rows x 8 columns]

In [ ]:
df[df['cluster'] == 3].sample(10) 

age  remote  benefits  tech  gender    supervisor  interfere  cluster
402    32       1         0     1    male           yes          3        3
882    33       0         0     0    male  some of them          1        3
1050   31       0         0     1    male           yes          0        3
818    31       1         0     1    male            no          2        3
1107   32       0         0     1    male           yes          1        3
382    31       0         0     1    male            no          0        3
47     33       0         0     1   other  some of them          0        3
939    32       0         1     0  female           yes          3        3
56     32       0         0     1  female  some of them          3        3
1084   31       0         1     1    male  some of them          3        3

### Systematic

Systematic Sampling is defined as the type of Probability Sampling where a researcher can research on a targeted data from large set of data. Targeted data is chosen by selecting random starting point and from that after certain interval next element is chosen for sample. In this a small subset (sample) is extracted from large data.

You can read more about systematic sampling [here](https://www.geeksforgeeks.org/systematic-sampling-in-pandas/).

For an example using our current dataset, suppose we sample from only those who work remote.

In [ ]:
remote_sample = df[df['remote'] == 1].sample(10)

In [ ]:
remote_sample

age  remote  benefits  tech  gender    supervisor  interfere  cluster
469    36       1         1     1    male            no          2        7
962    25       1         0     1  female            no          4        1
1232   39       1         1     1    male           yes          2        0
763    26       1         0     1    male           yes          2        1
193    39       1         0     1    male            no          1        0
603    38       1         0     0    male            no          2        0
838    27       1         0     1    male            no          1        1
253    35       1         0     0    male            no          1        7
61     41       1         1     1  female  some of them          2       10
1094   45       1         0     1    male  some of them          1        6

### Graph Sampling

Sometimes we come across data that needs to be sampled as a graph - we've seen earlier how we can also represent tables as graphs and vice-versa, so this should not be news to you. In this section we will explore some random ways of sampling from graphs. To illustrate these examples, we will be sampling from a graph based dataset. It is possible to extend these methods to your data that might not be (at first) represented in a graph-like way by restructuring your data in a way that supports these methods.

We will be using a python package that we briefly touched upon earlier, [Little Ball of Fur](https://arxiv.org/pdf/2006.04311.pdf).

#### Data

We will be using a dataset based on GitHub data. In this graph nodes represent GitHub developers and the edges between them are mutual follower relationships. For details about the dataset see this [paper](https://arxiv.org/abs/1909.13021Z).

In [ ]:
!pip install littleballoffur

     |████████████████████████████████| 8.4MB 6.4MB/s 
     |████████████████████████████████| 3.1MB 38.8MB/s 
  Created wheel for littleballoffur: filename=littleballoffur-2.1.8-cp37-none-any.whl size=40212 sha256=d6e3cd9136de854a67b5f0909d27f0efb2d3e713ad63f10d0e44037e2221d5a7
  Stored in directory: /root/.cache/pip/wheels/3d/09/b2/b2a80c4b1d51cc7de09f8de267455bcb40cd0b43c19dc79fa7
  Created wheel for networkit: filename=networkit-7.1-cp37-cp37m-linux_x86_64.whl size=8043401 sha256=b68e3b9d70f3e5d68fa934276a58935c1cd2727015a3d8118012d50014142ee7
  Stored in directory: /root/.cache/pip/wheels/a2/ca/1c/a1096b1823f845c29784ce5106866b01eaad26d101a79032b8
Successfully built littleballoffur networkit
ERROR: scikit-network 0.23.1 has requirement numpy>=1.20.2, but you'll have numpy 1.19.5 which is incompatible.
ERROR: scikit-network 0.23.1 has requirement scipy>=1.6.2, but you'll have scipy 1.4.1 which is incompatible.


In [ ]:
from littleballoffur import GraphReader

reader = GraphReader("github")

graph = reader.get_graph()

#### Node Sampling

We will first look at some popular node sampling algorithms. Let’s use the PageRank Proportional Node Sampling method from [Sampling From Large Graphs](https://cs.stanford.edu/people/jure/pubs/sampling-kdd06.pdf). We will sample approximately 50% of the original nodes from the network.

In [ ]:
from littleballoffur import PageRankBasedSampler, RandomNodeSampler


In [ ]:
number_of_nodes = int(0.5*graph.number_of_nodes())

In [ ]:
pagerank_sampler = PageRankBasedSampler(number_of_nodes = number_of_nodes)


In [ ]:
randomnode_sampler = RandomNodeSampler(number_of_nodes = number_of_nodes)

In [ ]:
randomnodes_graph = pagerank_sampler.sample(graph)

In [ ]:
pagerank_graph = pagerank_sampler.sample(graph)

#### Sub-graph Sampling

We will look at a series of exploration sampling algorithms that sample sub-graphs from larger graphs.

First is an implementation of node sampling by random walks. A simple random walker which creates an induced subgraph by walking around. 

In [ ]:
from littleballoffur import RandomWalkSampler

In [ ]:
randomwalk_sampler = RandomWalkSampler(number_of_nodes=number_of_nodes)

In [26]:
randomwalk_graph = randomnode_sampler.sample(graph)

Now let’s use the Metropolis-Hastings Random Walk Sampler method from [Metropolis Algorithms for Representative Subgraph Sampling](https://ieeexplore.ieee.org/document/4781123).  The random walker has a probabilistic acceptance condition for adding new nodes to the sampled node set. This constraint can be parametrized by the rejection constraint exponent. 

In [23]:
from littleballoffur import MetropolisHastingsRandomWalkSampler

In [24]:
mhrw_sampler = MetropolisHastingsRandomWalkSampler(number_of_nodes = number_of_nodes)


In [27]:
mhrw_graph = mhrw_sampler.sample(graph)

You will find many other sub-graph sampling algorithms, and many variations of the Random Walk algorithm (RandomWalkWithJumpSampler, RandomNodeNeighborSampler, RandomWalkWithRestartSampler) in the [Exploration Sampling](https://little-ball-of-fur.readthedocs.io/en/latest/modules/exploration_sampling.html). 


## Non-probabilistic Sampling and Extensions

A lot of the sampling we will see in this section involves sampling over a network or graph. For example, in a survey, we may ask one respondent to choose the next, and in that way create a sub-graph within the full social graph we are sampling from. 


### Snowball Sampling 

Snowball sampling is where research participants recruit other participants for a test or study. It is used where potential participants are hard to find. It’s called snowball sampling because (in theory) once you have the ball rolling, it picks up more “snow” along the way and becomes larger and larger. Snowball sampling is a non-probability sampling method. ([link to explanation](https://www.statisticshowto.com/probability-and-statistics/statistics-definitions/snowball-sampling/))

If we manually inspect the sampling, we can add our own criteria for stopping; in the package implementation, it stops when we have enough samples.


In [28]:
from littleballoffur import SnowBallSampler

In [29]:
snowball_sampler = SnowBallSampler(number_of_nodes = number_of_nodes)


In [30]:
snowball_graph = snowball_sampler.sample(graph)

littleballoffur also offers stochastic extensions of the Snowball Sampler. The Forest Fire Sampler is a stochastic snowball sampling method where the expansion is proportional to the burning probability, described [here](https://cs.stanford.edu/people/jure/pubs/sampling-kdd06.pdf). 

In [32]:
from littleballoffur import ForestFireSampler, SpikyBallSampler

In [33]:
forestfire_sampler = ForestFireSampler(number_of_nodes=number_of_nodes)

In [34]:
forestfire_graph = forestfire_sampler.sample(graph)

Below we use spiky ball sampling. The procedure is a filtered breadth-first search sampling method where the expansion is is performed over a random subset of neighbors. Originally described [here](https://www.mdpi.com/1999-4893/13/11/275).

In [35]:
spikyball_sampler = SpikyBallSampler(number_of_nodes=number_of_nodes)

In [36]:
spikyball_graph = spikyball_sampler.sample(graph)

That's the last of our graph based sampling. We will now discuss some other methods used. Note that the following sections are theoretical with no code, but the previous code can help you implement these methods.

### Convenience

This simply refers to using a sample that is at-hand and easy to analyse. For example, you may choose a free sub-graph of the full Facebook friend graph to run experiments because you do not have access to more data. You can refine your methods on this data before moving on to a larger dataset.

You can read more about it [here](https://methods.sagepub.com/reference/encyclopedia-of-survey-research-methods/n105.xml).

In [ ]:
# empty cell

### Quota Sampling

Very similar to stratified sampling, but we may not randonly sample from the strata and choose the whole sample. Here are some links to read more:

- [QuestionPro: Quota Sampling definition](https://www.questionpro.com/blog/quota-sampling/#:~:text=Quota%20sampling%20is%20defined%20as,to%20specific%20traits%20or%20qualities.)
- [Statistics How To: Quota Sampling](https://www.statisticshowto.com/quota-sampling/)
- [humansofdata: Quota Sampling](https://humansofdata.atlan.com/2016/04/quota-sampling-when-to-use-how-to-do-correctly/)


In [ ]:
# empty cell

### Purposive / Relevance / Judegement sampling

[text below from this link](https://www.alchemer.com/resources/blog/purposive-sampling-101/)

[paper: Purposeful sampling for qualitative data collection and analysis in mixed method implementation research](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4012002/)

Purposive sampling, also known as judgmental, selective, or subjective sampling, is a form of non-probability sampling in which researchers rely on their own judgment when choosing members of the population to participate in their surveys. 

This survey sampling method requires researchers to have prior knowledge about the purpose of their studies so that they can properly choose and approach eligible participants for surveys conducted.

Researchers use purposive sampling when they want to access a particular subset of people, as all participants of a survey are selected because they fit a particular profile. 

In [37]:
# empty cell

## Sampling Imbalanced Classes

A common problem we deal with in real world datasets is imbalance, when we have one (or a few) class, category or sections that have higher or lower representation than the other classes. This situation is often referred to as having imbalancing classes, and the python package [imbalanced-learn](https://github.com/scikit-learn-contrib/imbalanced-learn) is specifically built to tackle this.

In this section we will be following a tutorial by [Investigate.ai](https://investigate.ai/), a school for teaching data science to journalists. The tutorial we will follow is in this [repository](https://github.com/littlecolumns/ds4j-notebooks/blob/master/classification/notebooks/Correcting%20for%20imbalanced%20datasets.ipynb).



### Classification problems with imbalanced inputs

Oftentimes when we're doing real-world classification problems, we have the problem of **"imbalanced classes"**.

Let's say we're analyzing a document dump, and trying to find the documents that are interesting to us. Maybe we're only interested in 10% of them! The fact that there's such a bias - 90% of them are uninteresting - **will mess with our classifier.** Let's take a look at [imbalanced-learn](https://imbalanced-learn.readthedocs.io/en/stable/) library to help fix this problem!

### Prep work: Downloading necessary files
Before we get started, we need to download all of the data we'll be using.
* **recipes-indian.csv:** Indian classification recipes - a selection of recipe ingredient lists, with half of them being labeled as Indian cuisine
* **recipes.csv:** recipes - a selection of recipe ingredient lists, with each labeled with the cuisine its from


In [ ]:
# Make data directory if it doesn't exist
!mkdir -p data
!wget -nc https://nyc3.digitaloceanspaces.com/ml-files-distro/v1/classification/data/recipes-indian.csv -P data
!wget -nc https://nyc3.digitaloceanspaces.com/ml-files-distro/v1/classification/data/recipes.csv -P data

We're going to go through this pretty quickly, so you should be familiar with vectorizing, classification, and confusion matrices going in.

In [ ]:
import pandas as pd

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

### Our datasets

We're going to be looking at two datasets today. They're both **recipes and ingredient lists**, and with both we're predicting whether we can **accurate determine which recipes are Indian**.

Let's read them both in.

In [ ]:
df_balanced = pd.read_csv("data/recipes-indian.csv")
df_balanced['is_indian'] = (df_balanced.cuisine == "indian").astype(int)

df_balanced.head()

cuisine     id                                    ingredient_list  is_indian
0  indian  23348  minced ginger, garlic, oil, coriander powder, ...          1
1  indian  18869                           chicken, chicken breasts          1
2  indian  36405  flour, rose essence, frying oil, powdered milk...          1
3  indian  11494    soda, ghee, sugar, khoa, maida flour, milk, oil          1
4  indian  32675  tumeric, garam masala, salt, chicken, curry le...          1

In [ ]:
df_unbalanced = pd.read_csv("data/recipes.csv")
df_unbalanced['is_indian'] = (df_unbalanced.cuisine == "indian").astype(int)

df_unbalanced.head()

cuisine     id                                    ingredient_list  \
0        greek  10259  romaine lettuce, black olives, grape tomatoes,...   
1  southern_us  25693  plain flour, ground pepper, salt, tomatoes, gr...   
2     filipino  20130  eggs, pepper, salt, mayonaise, cooking oil, gr...   
3       indian  22213                  water, vegetable oil, wheat, salt   
4       indian  13162  black pepper, shallots, cornflour, cayenne pep...   

   is_indian  
0          0  
1          0  
2          0  
3          1  
4          1

They both look similar enough, right? A list of ingredients and an `is_indian` target column we'll be using as our label.

### Finding the imbalance

The real difference is how many of the recipes are Indian in each dataset. Let's take a look:

In [ ]:
df_balanced.is_indian.value_counts()

1    3000
0    3000
Name: is_indian, dtype: int64

In [ ]:
df_unbalanced.is_indian.value_counts()

0    36771
1     3003
Name: is_indian, dtype: int64

Ouch! That second dataset is really uneven - over ten times as many non-Indian recipes as there are Indian recipes!

The thing is: **this is usually how data looks in the real world.** You rarely have even numbers between your classes, and you often thing "more data is better data." We'll see how it plays out when we actually run our classifiers!

### Testing our datasets

We're going to use a `TfidfVectorizer` to convert ingredient lists to numbers, run a test/train split, and then train (and test) a `LinearSVC` classifier on the results. We'll start with the **balanced dataset**.

### Balanced dataset

In [ ]:
# Create a vectorizer and train it
vectorizer = TfidfVectorizer()
matrix = vectorizer.fit_transform(df_balanced.ingredient_list)

# Features are our matrix of tf-idf values
# labels are whether each recipe is Indian or not
X = matrix
y = df_balanced.is_indian

# How many are Indian?
y.value_counts()

1    3000
0    3000
Name: is_indian, dtype: int64

We still have an even split, 3000 non-Indian recipes and 3000 Indian recipes. Let's run a test/train split and see how the results look.

In [ ]:
# Split into test and train data
X_train, X_test, y_train, y_test = train_test_split(X, y)

# Build a classifier and train it
clf = LinearSVC()
clf.fit(X_train, y_train)

# Test our classifier and build a confusion matrix
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not indian', 'indian'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names).div(matrix.sum(axis=1), axis=0)

Predicted not indian  Predicted indian
Is not indian              0.962815          0.037185
Is indian                  0.048193          0.951807

**Our classifier looks pretty good!** Around 96% accuracy for predicting non-Indian food, and around 95% correctly predicting Indian food. High quality *and* even.

Let's move on to see how it looks with our **unabalanced dataset**.

### Unbalanced dataset 

In [ ]:
# Create a vectorizer and train it
vectorizer = TfidfVectorizer()
matrix = vectorizer.fit_transform(df_unbalanced.ingredient_list)

# Features are our matrix of tf-idf values
# labels are whether each recipe is Indian or not
X = matrix
y = df_unbalanced.is_indian

# How many are Indian?
y.value_counts()

0    36771
1     3003
Name: is_indian, dtype: int64

Again: around 36k non-Indian recipes really really outweighing the 3,003 Indian recipes. While we love the world of "more more more data," let's see what that imbalance does to our classifier.

In [ ]:
# Split our dataset is train and test data
X_train, X_test, y_train, y_test = train_test_split(X, y)

# Train the classifier on the training data
clf = LinearSVC()
clf.fit(X_train, y_train)

# Test our classifier and build a confusion matrix
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not indian', 'indian'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names).div(matrix.sum(axis=1), axis=0)

Predicted not indian  Predicted indian
Is not indian              0.992150          0.007850
Is indian                  0.180052          0.819948

Ouch!!! While we're doing **really well** at predicting non-Indian dishes, our ability to predict Indian dishes has plummeted to just over 80%.

Why does this happen? An easy way to think about it is **when it's a risky decision, it's always safest to guess "not Indian."** In fact, if we *always guessed non-Indian*, no matter what, we'd be right...

In [ ]:
36771/(36771+3003)

0.9244984160506864

About 92% of the time! So how do we solve this problem?

### Solving the problem

Solving the problem of unbalanced (or biased) input classes is actually not too hard! There's a nice library that can give us a hand, [imbalanced-learn](https://imbalanced-learn.readthedocs.io/en/stable/).

imbalanced-learn will **resample** our dataset, either generating new datapoints or pruning out existing datapoints, until the classes are evened out.

#### What do we resample?

An important thing to note is that **the problem with bias happens when we train our model.** If we show our model a skewed view of the world, it'll carry that bias when making judgments in the future. When we add or remove datapoints to even out the problem, **we only need to do this for the training data.**

We want to show the model an even view of the world, so we give it even data. The test data should still reflect the "real" world. Before we were looking at how imblanaced our overall dataset was, but now let's **just look at how biased the training data is.**

In [ ]:
y_train.value_counts()

0    27599
1     2231
Name: is_indian, dtype: int64

In [ ]:
y_train.value_counts(normalize=True)

0    0.92521
1    0.07479
Name: is_indian, dtype: float64

Looks like a little over 7% of our training data is Indian - we'd like to get that up to 50%, so let's see what the imbalanced-learn library can do for us!

#### Undersampling

If we're feeling guilty that there are so many additional non-Indian recipes, *we could always get rid of those extra non-Indian recipes!* In fact, the balanced dataset was me manually creating a new CSV from an even split of Indian/non-Indian recipes..

Instead of manually digging through our dataset to even things out, though, we can rely on imbalanced-learn to do it automatically. We'll use the technique of **undersampling** to take those ~28k non-Indian recipes and randomly filter them down to around 2,000 to match the number of Indian recipes. (Remember we're only doing this with training data!)

In [ ]:
from imblearn.under_sampling import RandomUnderSampler

resampler = RandomUnderSampler()
# Resample X and y so there are equal numbers of each y
X_train_resampled, y_train_resampled = resampler.fit_resample(X_train, y_train)

y_train_resampled.value_counts()

1    2231
0    2231
Name: is_indian, dtype: int64

Okay, cool, equal numbers! Let's see how the classifier performs.

In [ ]:
# We already split our data, so we don't need to do that again

# Train the classifier on the resampled training data
clf = LinearSVC()
clf.fit(X_train_resampled, y_train_resampled)

# Build a confusion matrix
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not indian', 'indian'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names).div(matrix.sum(axis=1), axis=0)

Predicted not indian  Predicted indian
Is not indian              0.957479          0.042521
Is indian                  0.051813          0.948187

Looking good! It performs as well as our other 3,000/3,000 split because, well, it's more or less the same thing (although the test data is "realistically" unbalanced).

#### Oversampling

Cutting out those 27,000 "extra" non-Indian recipes seems like such a bummer, though. Wouldn't it be nice if we somehow found another 25,000 Indian recipes to even up our unbalanced training dataset to 27k non-Indian and 27k Indian? It's possible with **oversampling!**

Oversampling generates **new datapoints** based on your existing dataset. In this case we're going to use the `RandomOverSampler`, which just fills our dataset with **copies of the less-included class**. We'll have 27k Indian recipes, *but they'll be 25,0000 copies of the original ones*. Can that possibly help?

In [ ]:
from imblearn.over_sampling import RandomOverSampler

resampler = RandomOverSampler()
X_train_resampled, y_train_resampled = resampler.fit_resample(X_train, y_train)

In [ ]:
y_train_resampled.value_counts()

1    27599
0    27599
Name: is_indian, dtype: int64

Looking good, a nice even 27,599 apiece. Let's see how the classifier works out!

In [ ]:
# We already split our dataset into train and test data

# Train the classifier on the resampled training data
clf = LinearSVC()
clf.fit(X_train_resampled, y_train_resampled)

# Build a confusion matrix with the result
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not indian', 'indian'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names).div(matrix.sum(axis=1), axis=0)

Predicted not indian  Predicted indian
Is not indian              0.969363          0.030637
Is indian                  0.068653          0.931347

Also looking pretty good! A little bit better at predicting non-Indian dishes and a little bit worse at predicting Indian dishes, but it more or less evens out with the undersampled example. 

There are also other oversampling techniques that involve **creating synthetic data,** new datapoints that aren't *copies* of our data, but rather totally new ones. You can read more about them [on the imbalanced-learn page](https://imbalanced-learn.readthedocs.io/en/stable/over_sampling.html).

#### Review

In this section we talked about the problem of **imbalanced classes**, where an uneven split in your labels can cause suboptimal classifier performance. We used the imbalanced-learn library to talk about two methods of solving the issue - undersampling and oversampling - which both boosted performance as compared to the imbalanced dataset.

#### Discussion topics

What is the difference between oversampling and undersampling? Why might have oversampling done a better job predicting non-Indian recipes?

Why did we only resample the training data, and not the test data?

While the idea of automatically-generated fake data might sound more attractive than just re-using existing data, [what might be some issues with it](https://imbalanced-learn.readthedocs.io/en/stable/over_sampling.html)?

Can we think of any times when we might *not* want a balanced dataset?

You can find more examples with imbalanced datasets in the [example gallery](https://imbalanced-learn.org/stable/auto_examples/index.html#general-examples).

## Bootstrap sampling and sub-sampling

Bootstrap sampling and sub-sampling are two methods that are particularly relevant in the context of machine learning and deep learning. 

[original text from this blog](https://machinelearningmastery.com/a-gentle-introduction-to-the-bootstrap-method/).

The bootstrap method is a resampling technique used to estimate statistics on a population by sampling a dataset **with replacement**.

It can be used to estimate summary statistics such as the mean or standard deviation. It is used in applied machine learning to estimate the skill of machine learning models when making predictions on data not included in the training data.

A desirable property of the results from estimating machine learning model skill is that the estimated skill can be presented with confidence intervals, a feature not readily available with other methods such as cross-validation.

### Bootstrap sampling example with word embeddings

In this section we will explore the stability of word embeddings using bootstrap sampling. This is a well researched question in NLP research ([Evaluating the Stability of Embedding-based Word Similarities](https://mimno.infosci.cornell.edu/papers/antoniak-stability.pdf)). In the paper, *the authors come to the conclusion that there are several sources of variability
in cosine similarities between word embeddings vectors. The size of the corpus, the length of individual documents, and the presence or absence of specific
documents can all affect the resulting embeddings. While differences in word association are measurable and are often significant, small differences in cosine similarity are not reliable, especially for small corpora. If the intention of a study is to learn about a specific corpus, they recommend that practitioners test the statistical confidence of similarities based on
word embeddings by training on multiple bootstrap samples.*

We will conduct a mini-version of this experiment to test how stable word similarities are after training with and without bootrstrap sampling on a smaller text dataset. We will use a dataset we used before, the hobbies dataset.

In [38]:
import gensim

In [39]:
from yellowbrick.datasets import load_hobbies

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.metrics.classification module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.metrics. Anything that cannot be imported from sklearn.metrics is now part of the private API.
  warnings.warn(message, FutureWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.datasets.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.datasets. Anything that cannot be imported from sklearn.datasets is now part of the private API.
  warnings.warn(message, FutureWarning)


In [40]:
from gensim.parsing.preprocessing import preprocess_documents

In [41]:
corpus = load_hobbies()

In [42]:
preprocessed_texts = preprocess_documents(corpus.data)

In [52]:
len(preprocessed_texts)

448

In [45]:
from gensim.models import Word2Vec


In [47]:
w2vmodel_cleaned = Word2Vec(
        preprocessed_texts,
        size=100,
        window=10)

In [48]:
w2vmodel_cleaned.wv.most_similar("book")

[('read', 0.9998674392700195),
 ('stori', 0.9998670816421509),
 ('publish', 0.9998620748519897),
 ('author', 0.9998588562011719),
 ('movi', 0.9998471140861511),
 ('edit', 0.9998458027839661),
 ('futur', 0.9998434782028198),
 ('turn', 0.9998424649238586),
 ('bui', 0.9998399019241333),
 ('page', 0.9998395442962646)]

We will use this as a reference while we try out how stable the model is with two sets of bootstrapped texts!

In [49]:
from sklearn.utils import resample

In [50]:
bootstrap_texts = resample(preprocessed_texts)

In [51]:
len(bootstrap_texts)

448

In [54]:
w2vmodel_boot = Word2Vec(
        bootstrap_texts,
        size=100,
        window=10)

In [55]:
w2vmodel_boot.wv.most_similar("book")

[('author', 0.9998607635498047),
 ('publish', 0.9998499155044556),
 ('broken', 0.9997895956039429),
 ('compani', 0.9997869729995728),
 ('featur', 0.9997847080230713),
 ('contain', 0.9997791051864624),
 ('stori', 0.9997789859771729),
 ('page', 0.9997786283493042),
 ('futur', 0.9997777938842773),
 ('novel', 0.9997714757919312)]

In [56]:
bootstrap_texts_1 = resample(preprocessed_texts)

In [57]:
w2vmodel_boot_1 = Word2Vec(
        bootstrap_texts_1,
        size=100,
        window=10)

In [58]:
w2vmodel_boot_1.wv.most_similar("book")

[('stori', 0.9997271299362183),
 ('femal', 0.9997173547744751),
 ('write', 0.9996861815452576),
 ('read', 0.9996781349182129),
 ('actor', 0.9996703863143921),
 ('compani', 0.9996675252914429),
 ('batman', 0.9996577501296997),
 ('present', 0.9996563196182251),
 ('product', 0.9996486902236938),
 ('futur', 0.9996473789215088)]

Smaller corpora are likely to be more variable in their embeddings, and when bootstrapped we see this more clearly. We encourage you to read the paper on stability in word embeddings and to run your own experiments with your datasets.

### Sub-sampling

Unlike bootstrap sampling, in sub-sampling we sample without replacement, usually in cases when we are dealing with very large data and only want a portion of it. [subsample](https://pypi.org/project/subsample/) is a python package that works as a command-line tool for sub-sampling, and is especially powerful with text based datasets. It includes methods such as reservoir sampling and two pass sampling. 

# Annotation & Reliability

Up until this week, we have assumed that the corpus you have used for analysis assignments represented a *meaningful* assemblage of texts from which reasonable inferences could be drawn about the social game, social world and social actors that produced it. 

This week, we ask you to build a corpus for preliminary analysis and articulate what your sample represents in context of your final project. We begin by exploring how we can get *human* readings of content at scale. We want to gather and utilize human responses for several reasons. First, we may want to use crowdsourced human scores as the primary method of coding, extracting or organizing content (as it was in two of the assigned readings). Second, we may want to validate or tune a computational algorithm we may have developed in terms of how it is associated with human meanings or experience. Finally, we may want to use human coding on a sample of data as the basis for training a model or algorithm to then extrapolate *human-like* annotations to the entire population. Here intelligent sampling is critical to maximize effective maching training. 

For this notebook we will be using the following packages

In [ ]:
! pip install -U git+git://github.com/UChicago-Computational-Content-Analysis/lucem_illud.git

  Cloning git://github.com/UChicago-Computational-Content-Analysis/lucem_illud.git to /tmp/pip-req-build-sydlad_r
  Running command git clone -q git://github.com/UChicago-Computational-Content-Analysis/lucem_illud.git /tmp/pip-req-build-sydlad_r
     |████████████████████████████████| 5.6MB 8.5MB/s 
     |████████████████████████████████| 122kB 49.2MB/s 
     |████████████████████████████████| 174kB 52.1MB/s 
     |████████████████████████████████| 81kB 13.5MB/s 
     |████████████████████████████████| 133kB 46.4MB/s 
     |████████████████████████████████| 32.8MB 202kB/s 
     |████████████████████████████████| 71kB 13.1MB/s 
     |████████████████████████████████| 5.1MB 19.9MB/s 
     |████████████████████████████████| 81kB 12.3MB/s 
     |████████████████████████████████| 7.6MB 35.0MB/s 
  Created wheel for lucem-illud: filename=lucem_illud-8.0.1-cp37-none-any.whl size=34956 sha256=288546caa261f0ef241d83d00740217b5bf49e5afae67f096ea2f4fe8ab8a3c0
  Stored in directory: /tmp/pip-ephem

In [ ]:
#This provides access to data and to helper functions from previous weeks
#Make sure you update it before starting this notebook
import lucem_illud #pip install -U git+git://github.com/UChicago-Computational-Content-Analysis/lucem_illud.git

#All these packages need to be installed from pip
import numpy as np #For arrays
import scipy as sp #For some stats
import pandas #Gives us DataFrames
import matplotlib.pyplot as plt #For graphics
import seaborn #Makes the graphics look nicer
import pyanno #On python3 make sure to pip install pyanno3

#We need to import these this way due to how pyanno is setup
from pyanno.measures import pairwise_matrix, agreement, cohens_kappa, cohens_weighted_kappa, fleiss_kappa, krippendorffs_alpha, pearsons_rho, scotts_pi, spearmans_rho
from pyanno.annotations import AnnotationsContainer
from pyanno.models import ModelA, ModelBt, ModelB

from functools import reduce
from itertools import permutations
import math


#This 'magic' command makes the plots work better
#in the notebook, don't use it outside of a notebook.
#Also you can ignore the warning
%matplotlib inline

import os #For looking through files
import os.path #For managing file paths

## Example Annotation Dataset

Load Rzhetsky et al (2009)'s sample dataset, which can be found [here](https://github.com/enthought/uchicago-pyanno/tree/master/data). This data is the result of a content analytic / content extraction study in which Andrey Rzhetsky and colleagues from the National Library of Medicine, published [here](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000391) in [PLOS Computational Biology](http://journals.plos.org/ploscompbiol/), gave eight annotators 10,000 sentence chunks from biomedical text in biomedical abstracts and articles, then asked them, in a loop design schematically illustrated below that provided 3 independent codings for each document. The sampling strategy pursued diversity by drawing from PubMed abstracts (1000) and full-text articles (9000: 20% from abstracts, 10% from introductions, 20% from methods, 25% from results, and 25% from discussions.) The dataset extract here involves respondents codes for sentences in terms of their *Evidence*: {0, 1, 2, 3, -1} where 0 is the complete lack of evidence, 3 is direct evidence present within the sentence, and -1 is didn't respond. (They also crowdsourced and analyzed *polarity*, *certainty*, and *number*). For example, consider the following two abutting sentence chunks: *"Because null mutations in toxR and toxT abolish CT and TcpA expression in the El Tor biotype and also attenuate virulence..."* [i.e., average certainty = 0], *"...it is likely that the ToxR regulon has functional similarities between the two biotypes despite the clear differences in the inducing parameters observed in vitro"* [i.e., average certainty = 1]."

In [ ]:
%%html
<img source="loopdesign.png">

[Click here for loop design](loopdesign.png)

In [ ]:
x = np.loadtxt("/content/pyAnno/testdata_numerical.txt")
anno = AnnotationsContainer.from_array(x, missing_values=[-1])

Interrogate the AnnotationsContainer object.

In [ ]:
anno.annotations

In [ ]:
anno.labels

In [ ]:
anno.missing_values

## Annotation Statistics

First, we assume categorical codes...that each code is qualitatively distinct from each other. Two measures are primarily used for this: Scott's $\pi$, Cohen's $\kappa$, and Krippendorff's $\alpha$ which each measure the extent of agreement between two annotators, but take into account the possibility of the agreement occurring by chance in slightly different ways. Any agreement measure begins with the frequency of codes:

In [ ]:
pyanno.measures.agreement.labels_frequency(anno.annotations,4)

Now consider the "confusion matrix" or matrix of coded agreements between any two coders:

In [ ]:
c = pyanno.measures.agreement.confusion_matrix(anno.annotations[:,0], anno.annotations[:,1],4)
print(c)
ac = seaborn.heatmap(c)
plt.show()

Scott's $\pi$ is computed as:

$\pi = \frac{\text{Pr}(a)-\text{Pr}(e)}{1-\text{Pr}(e)}$

Where Pr($a$) is relative observed agreement, and Pr($e$) is expected agreement using joint proportions calculated from the confusion matrix or matrix of coded agreements between any two coders:

In [ ]:
scotts_pi(anno.annotations[:,0], anno.annotations[:,1])

The generalization of Scott's $\pi$ to $n$ coders is Fleiss' $\kappa$ (Fleiss called it $\kappa$ because he thought he was generalizing Cohen's $\kappa$)

In [ ]:
fleiss_kappa(anno.annotations[::])

Krippendorff's $\alpha$ generalizes of Fleiss' $\kappa$ to $n$ coders and takes into account the fact that annotations here are not categorically different, but ordinal, by adding a weight matrix in which off-diagonal cells contain weights indicating the seriousness of the disagreement between each score. When produced with no arguments, it simply produces an arithmetic distance (e.g., 3-1=2), such that cells one off the diagonal are weighted 1, two off 2, etc.

In [ ]:
krippendorffs_alpha(anno.annotations[::])

Like Scott's $\pi$, Cohen's $\kappa$ also takes into account the possibility of the agreement occurring by chance, but in the following way:

$\kappa = \frac{p_o-p_e}{1-p_e}=1-\frac{1-p_o}{p_e}$

where $p_o$ is the relative observed agreement among raters, and $p_e$ is the hypothetical probability of chance agreement, using the observed data to calculate the probabilities of each observer randomly saying each category. If the raters are in complete agreement then $\kappa = 1$. If there is no agreement among the raters other than what would be expected by chance (as given by $p_e$), $\kappa ≤ 0 $. Here, Cohen's $\kappa$ statistic for the first two annotators is computed. This is probably the most common metric of agreement.

In [ ]:
cohens_kappa(anno.annotations[:,0], anno.annotations[:,1])

In [ ]:
m = pairwise_matrix(cohens_kappa, anno.annotations)
print(m)

In [ ]:
ax = seaborn.heatmap(m)
plt.show()

You can see that this 8 by 3 loop design will be less stable than an 8 choose 3 combinatorial design, because each codes with more others. 

One can also assess the average Cohen's $\kappa$ for all pairs of coders that have coded against one another:

In [ ]:
def pairwise_metric_average(metric, array):
    """Calculate the pairwise metric average for the real elements of metric function run on an array of annotations"""
    p = permutations(range(array[0,:].size),2)
    m = [metric(array[:,x[0]], array[:,x[1]]) for x in p]
    clean_m = [c for c in m if not math.isnan(c)]
    return reduce(lambda a, b: a + b, clean_m)/len(clean_m)    
 
pairwise_metric_average(cohens_kappa, anno.annotations)

As recognized with Krippendorff's flexible $\alpha$, our scores are *not* categorical, but rather ordered and her considered metric. Weighted $\kappa$ allows you to count disagreements differently and is useful when codes are ordered as they are here. Here a weight matrix is added to the calculation, in which off-diagonal cells contain weights indicating the seriousness of the disagreement between each score. When automatically produced, it simply produces an arithmetic distance (e.g., 3-1=2), such that cells one off the diagonal are weighted 1, two off 2, etc. Here

$\kappa = 1-\frac{\sum^k_{i=1}\sum^k_{j=1}w_{ij}x_{ij}}{\sum^k_{i=1}\sum^k_{j=1}w_{ij}m_{ij}}$

where $\kappa$ = $n$ codes and $w_{ij}$,$x_{ij}$, and $m_{ij}$ represent elements in the weight, observed, and expected matrices, respectively. (Obviously, when diagonal cells contain weights of 0 and off-diagonal cells weights of 1, this equals $\kappa$).

In [ ]:
cohens_weighted_kappa(anno.annotations[:,0], anno.annotations[:,1])

Or averaged over the total:

In [ ]:
pairwise_metric_average(cohens_weighted_kappa,anno.annotations)

Alternatively, if the annontation data can be understood as indicating real values, we can assess not agreement, but rather the correlation of values (Pearson's $\rho$) or correlation of ranks (Spearman's $\rho$) for pairs of coders:

In [ ]:
n = pairwise_matrix(pearsons_rho, anno.annotations)
m = pairwise_matrix(spearmans_rho, anno.annotations)
an = seaborn.heatmap(n)
plt.show()
am = seaborn.heatmap(m)
plt.show()

Or averaged over all comparable pairs:

In [ ]:
print(pairwise_metric_average(pearsons_rho,anno.annotations), pairwise_metric_average(spearmans_rho,anno.annotations))

## Models

However, what if some coders are better than others. The prior measures all rely on the assumption that all coders are equally good. What if some are worse than others? Now we use Rzhetsky et al (2009) and Dawid & Skene's models to make inference about true label classes by downweighting bad or deviant coders. Pyanno provides three relevant models: ModelA, ModelB, and ModelBt. Model A can only be currently run on a balanced 8-coder design, but assesses accuracy purely based on agreement. Model B with $\theta$s models the relationship between each coder and code. Model B is the Dawid & Skene model from the reading. The following image schematically suggests the relationship between the models. <img src="../data/models.png">

The models should provide similar results. To estimate the parameters for any models, we first need to create a new model. 

In [ ]:
# create a new instance of model A, with 4 label classes
model = ModelB.create_initial_state(4, 8)
# other model parameters are initialized from the model prior
print(model.theta)
print(model.log_likelihood(anno.annotations))

In [ ]:
samples = model.sample_posterior_over_accuracy(anno.annotations, 200, burn_in_samples=100, thin_samples=3)

Pyanno allows one to use either MLE (maximum likelihood estimation) or MAP (maximum a posteriori estimation) to estimate model parameters. Note that the parameters here correspond to our estimation of the accuracy of each annotator.

In [ ]:
model.map(anno.annotations)
print(model.theta)
print(model.log_likelihood(anno.annotations))

In [ ]:
model = ModelB.create_initial_state(4, 8)
model.map(anno.annotations)
print(model.theta)
print(model.log_likelihood(anno.annotations))

Once we have model parameters estimated, we can now make inferences about the true label classes. We can calculate the posterior distribution over the true label classes.

In [ ]:
posterior = model.infer_labels(anno.annotations)
print(posterior)

Let's turn the posterior of the first 100 samples into a heatmap.

In [ ]:
votes = []
for r in anno.annotations:
    v = [0] * len(anno.labels)
    votes.append(v)
    for a in r:
        if a > -1:
            v[a] += 1
votes_array = np.array(votes)

In [ ]:
fig, (ax1, ax2) = plt.subplots(ncols=2, figsize = (15, 10), sharey=True)
num_questions = 20

seaborn.heatmap(votes_array[:num_questions], annot = True, ax=ax2)
seaborn.heatmap(posterior[:num_questions], annot=True, ax =ax1)
ax1.set_title("Model")
ax2.set_title("Votes")
plt.show()

This differs markedly from taking annotator scores at face value (Add comparison of average values)

In [ ]:
samples = model.sample_posterior_over_accuracy(anno.annotations, 200, burn_in_samples=100, thin_samples=3)

In [ ]:
print(samples[0].mean(axis=0))
print(samples[0].std(axis=0))

Let's try everything again with ModelBt

In [ ]:
# create a new instance of model B, with 4 label classes and 8 annotators.
model = ModelBt.create_initial_state(4, 8)
print(model.theta)
print(model.log_likelihood(anno.annotations))

In [ ]:
model.map(anno.annotations)
print(model.theta)
print(model.log_likelihood(anno.annotations))

In [ ]:
posterior = model.infer_labels(anno.annotations)
print(posterior)

Let's visualize the posterior of the first 10 samples according to ModelBt.

In [ ]:
ax = seaborn.heatmap(posterior[:10,])
plt.show()

The property of these scores is that they enable us to identify the most likely code assuming coders of unequal quality, which also allows us to break ties when we know coder identity. For some analyses, we may simply use the posterior themselves rather than the most probably code outcome.

## Generating Annotations

Pyanno also allows one to generate artificial data from a model.

In [ ]:
model = ModelBt.create_initial_state(4, 3, theta=[0.99,0.75,0.25])
#randome generate annotations with 4 label classes and 3 annotators. The accuracy of the three annotators are 0.99, 0.75, and 0.25 respectively.
model.generate_annotations(20)

## Visualizing coder accuracy

Pyanno provides a [graphical user interface](http://docs.enthought.com/uchicago-pyanno/user_guide.html) for making plots. However, it is not compatible with ipython notebooks. Nevertheless, nothing prevents us from making plots using matplotlib. Let's make a plot of the accuracy of each annotator inferred from ModelA.

In [ ]:
model = ModelBt.create_initial_state(4, 8)
model.mle(anno.annotations)
samples = model.sample_posterior_over_accuracy(anno.annotations, 200, burn_in_samples=100, thin_samples=3)
y =  samples.mean(axis=0)#.mean(axis = 1).mean(axis = 1)
y_ci = samples.std(axis=0)#.mean(axis = 1).mean(axis = 1)

In [ ]:
plt.figure()
plt.errorbar(range(8),y, yerr = y_ci)
plt.show()

## Example with articles that use the General Social Survey

I performed a recent study in which the variables from thousands of articles were associated with those used in the General Social Survey, a widely used population sample, in order to interrogate how social science analyses are performed. Each article was reread and coded by a balanced set of three student coders using a 6 choose 3 design, such that all possible 3-coder-subsets (20) coded an equal number of articles. Coding was performed through a website that allowed students access to the digital article. To evaluate the validity of the student codes, we also recruited a sample of authors associated with 97 of our published articles to fill out the same online survey. 

Because not all coders coded items with equal accuracy, and because “don’t know” was an optional answer, leading to potential ties, we used a generative, probabilistic model to estimate the maximum a posteriori probability (MAP) prediction that an item’s code is true, which integrates over the estimated accuracy of coders, assuming only that the entire population of coders is slightly more often right than wrong. The model (“Model B”) is based on a simple underlying generation process that directly accounts for the probability that coded values are correct (Rzhetsky et al. 2009). For each coded value j, a set of parameters, denoted γj, represents the probability that each coded value is correct. For the ith coder (i = 1, 2, …, 6), we introduce a matrix of probabilities, denoted λ(i)x|y, that defines the probability that she assigns code x (e.g., Dependent variable) to a GSS variable with correct annotation y. For a perfect coder, the matrix λ(i)x|y would equal the identity matrix and her vote would count most toward the total. For a coder that always codes incorrectly—a “troll”—her matrix λ(i)x|y will have all its value off the diagonal and will only minimally influence the posterior. We co-authored the open source pyanno software that implements this model.

Getting the data for each content analysis survey regarding how GSS variables were used in a large population of social science articles.

In [ ]:
#anno_vdep = AnnotationsContainer.from_file(missing_values=[-1], filename="GSSvariable_testSdependent.csv")
dev = np.loadtxt(fname="/content/dataforgssstudy/n7GSSvariable_testSdependent.csv", dtype=int, delimiter=",")
anno_dv = AnnotationsContainer.from_array(dev)

ind = np.loadtxt(fname="/content/dataforgssstudy/n7GSSvariable_testSindependent.csv", dtype=int, delimiter=",")
anno_iv = AnnotationsContainer.from_array(ind)

cent = np.loadtxt(fname="/content/dataforgssstudy/n7GSSvariable_testScentral.csv", dtype=int, delimiter=",")
anno_cv = AnnotationsContainer.from_array(cent)

cont = np.loadtxt(fname="/content/dataforgssstudy/n7GSSvariable_testScontrol.csv", dtype=int, delimiter=",")
anno_ctv = AnnotationsContainer.from_array(cont)

test = np.loadtxt(fname="/content/dataforgssstudy/testH.csv", dtype=int, delimiter=",")
anno_test = AnnotationsContainer.from_array(test)

Let's examine the data structure.

In [ ]:
dev.shape

In [ ]:
anno_dv.labels

In [ ]:
anno_dv.missing_values

In [ ]:
anno_dv.annotations.shape

First, let's use Cohen's $\kappa$ to measure agreement between coders...

In [ ]:
m = pairwise_matrix(cohens_kappa, anno_dv.annotations)
print(m)

Let's visualize that...

In [ ]:
ax = seaborn.heatmap(m)
plt.show()

In [ ]:
pairwise_metric_average(cohens_kappa, anno_dv.annotations)

Let's compute the statistics on each of the datasets and with Pearson's $\rho$. 

In [ ]:
datasets = [anno_dv.annotations, anno_iv.annotations, anno_cv.annotations, anno_ctv.annotations]
ck = [pairwise_matrix(cohens_kappa, anno) for anno in datasets]
pr = [pairwise_matrix(pearsons_rho, anno) for anno in datasets]
titles = ['DV', 'IV', 'Central Variable', "Control Variable"]

In [ ]:
fig, axs = plt.subplots(2,4)
fig.set_size_inches(18, 7)
for k, ax, title in zip(ck,axs[0], titles):
    seaborn.heatmap(k, ax = ax)
    ax.set_title(title)
    ax.set_xticks(())
    ax.set_yticks(())
for r, ax in zip(pr,axs[1]):
    seaborn.heatmap(r, ax = ax)
    ax.set_xticks(())
    ax.set_yticks(())
plt.show()

Now we will compare the student coders.

In [ ]:
nondiag = (np.eye(6)-np.ones(6))*-1.0

In [ ]:
xdevck = pairwise_matrix(cohens_kappa, anno_dv.annotations)
xdevpr = pairwise_matrix(pearsons_rho, anno_dv.annotations)

xindck = pairwise_matrix(cohens_kappa, anno_iv.annotations)
xindpr = pairwise_matrix(pearsons_rho, anno_iv.annotations)

xcenck = pairwise_matrix(cohens_kappa, anno_cv.annotations)
xcenpr = pairwise_matrix(pearsons_rho, anno_cv.annotations)

xconck = pairwise_matrix(cohens_kappa, anno_ctv.annotations)
xconpr = pairwise_matrix(pearsons_rho, anno_ctv.annotations)

print(np.average(xdevck, weights=nondiag))
print(np.average(xdevpr, weights=nondiag))
print(np.average(xindck, weights=nondiag))
print(np.average(xindpr, weights=nondiag))
print(np.average(xcenck, weights=nondiag))
print(np.average(xcenpr, weights=nondiag))
print(np.average(xconck, weights=nondiag))
print(np.average(xconpr, weights=nondiag))

Now we are going to bring in "gold standard" data. In this case, this is where we asked authors of the articles to code their own article's variables and compare with our student coders.

In [ ]:
mergedata = np.loadtxt(fname="/content/dataforgssstudy/gss_mergedataC.txt", dtype=int, delimiter="\t")

In [ ]:
anno_merge_dep = AnnotationsContainer.from_array(mergedata[:,0:2])
anno_merge_ind = AnnotationsContainer.from_array(mergedata[:,2:4])
anno_merge_cen = AnnotationsContainer.from_array(mergedata[:,4:6])
anno_merge_con = AnnotationsContainer.from_array(mergedata[:,6:8])
anno_merge_dkn = AnnotationsContainer.from_array(mergedata[:,8:10])

In [ ]:
print("""Dependent variable -- kappa & rho""")
print(cohens_kappa(anno_merge_dep.annotations[:,0], anno_merge_dep.annotations[:,1]))
print(pearsons_rho(anno_merge_dep.annotations[:,0], anno_merge_dep.annotations[:,1]))

print("\nIndependent variable")
print(cohens_kappa(anno_merge_ind.annotations[:,0], anno_merge_ind.annotations[:,1]))
print(pearsons_rho(anno_merge_ind.annotations[:,0], anno_merge_ind.annotations[:,1]))

print("\nCentral variable")
print(cohens_kappa(anno_merge_cen.annotations[:,0], anno_merge_cen.annotations[:,1]))
print(pearsons_rho(anno_merge_cen.annotations[:,0], anno_merge_cen.annotations[:,1]))

print("\nControl variable")
print(cohens_kappa(anno_merge_con.annotations[:,0], anno_merge_con.annotations[:,1]))
print(pearsons_rho(anno_merge_con.annotations[:,0], anno_merge_con.annotations[:,1]))

Whoah! Student coders and authors viewed articles that were "central" or critical to the published argument as fundamentally different (exhibiting negative agreement and correlation). Why? Likely because that researchers recalled what they had _intended_ as their central variables before analysis, but those that _worked out_ became central in the text.

Now for the assessment of the relative values of authors, then student coders.

In [ ]:
print("Dependent")
print(np.average(anno_merge_dep.annotations[:,0]))
print(np.average(anno_merge_dep.annotations[:,1]))

print("\nIndependent")
print(np.average(anno_merge_ind.annotations[:,0]))
print(np.average(anno_merge_ind.annotations[:,1]))

print("\nCentral")
print(np.average(anno_merge_cen.annotations[:,0]))
print(np.average(anno_merge_cen.annotations[:,1]))

print("\nControl")
print(np.average(anno_merge_con.annotations[:,0]))
print(np.average(anno_merge_con.annotations[:,1]))

## Now we are going to use models to predict the correct annotations

Recall that Model A is built for 8 coders, but we have 6. We're going to *hack* it by adding two blank columns.

In [ ]:
dev.shape

In [ ]:
negs2 = np.ones((21461, 2), dtype=np.int)*(-1)
devA = np.concatenate((dev, negs2), axis=1)
devA

In [ ]:
anno_dvA = AnnotationsContainer.from_array(devA)
model_devA = ModelA.create_initial_state(2)
model_devA.theta

In [ ]:
model_dvB = ModelB.create_initial_state(2, 6)
print(model_dvB.pi)
print(model_dvB.log_likelihood(anno_dv.annotations))

In [ ]:
model_dvB.map(anno_dv.annotations)
print(model_dvB.pi)
print(model_dvB.log_likelihood(anno_dv.annotations))

In [ ]:
# compute the posterior distribution over true annotations
posterior_dvB = model_dvB.infer_labels(anno_dv.annotations)
# each row show the probability of each label class for the
# corresponding item
print(posterior)

In [ ]:
samples_dvB = model_dvB.sample_posterior_over_accuracy(anno_dv.annotations, 200, burn_in_samples=100, thin_samples=3)

In [ ]:
# we can then compute a credible interval for the parameters:
ci_dv_mean = samples_dvB[0].mean(axis=0)
print("Mean")
print(ci_dv_mean)

ci_dv_stdev = samples_dvB[0].std(axis=0)
print("\nSTD")
print(ci_dv_stdev)


We will use Model B estimates for other variable assessments.

In [ ]:
#test
model_testB = ModelB.create_initial_state(2, 6)
print(model_testB.log_likelihood(anno_test.annotations))
model_testB.map(anno_test.annotations)
print(model_testB.pi)
print(model_testB.log_likelihood(anno_test.annotations))
print(anno_test.annotations.shape)
posterior_testB = model_testB.infer_labels(anno_test.annotations)
print(posterior_testB.shape)
samples_testB = model_testB.sample_posterior_over_accuracy(anno_test.annotations, 200, burn_in_samples=100, thin_samples=3)
ci_test_mean = samples_testB[0].mean(axis=0)
print(ci_test_mean)

In [ ]:
#indepedent variables
model_ivB = ModelB.create_initial_state(2, 6)
print(model_ivB.log_likelihood(anno_iv.annotations))
model_ivB.map(anno_iv.annotations)
print(model_ivB.pi)
print(model_ivB.log_likelihood(anno_iv.annotations))
print(anno_iv.annotations.shape)
posterior_ivB = model_ivB.infer_labels(anno_iv.annotations)
print(posterior_ivB.shape)
samples_ivB = model_ivB.sample_posterior_over_accuracy(anno_iv.annotations, 200, burn_in_samples=100, thin_samples=3)
ci_iv_mean = samples_ivB[0].mean(axis=0)
print(ci_iv_mean)

#central variables
model_cvB = ModelB.create_initial_state(2, 6)
print(model_cvB.log_likelihood(anno_cv.annotations))
model_cvB.map(anno_cv.annotations)
print(model_cvB.pi)
print(model_cvB.log_likelihood(anno_cv.annotations))
print(anno_cv.annotations.shape)
posterior_cvB = model_cvB.infer_labels(anno_cv.annotations)
print(posterior_cvB.shape)
samples_cvB = model_cvB.sample_posterior_over_accuracy(anno_cv.annotations, 200, burn_in_samples=100, thin_samples=3)
ci_cv_mean = samples_cvB[0].mean(axis=0)
print(ci_cv_mean)

#control variables
model_ctvB = ModelB.create_initial_state(2, 6)
print(model_ctvB.log_likelihood(anno_ctv.annotations))
model_ctvB.map(anno_ctv.annotations)
print(model_ctvB.pi)
print(model_ctvB.log_likelihood(anno_ctv.annotations))
print(anno_ctv.annotations.shape)
posterior_ctvB = model_ctvB.infer_labels(anno_ctv.annotations)
print(posterior_ctvB.shape)
samples_ctvB = model_ctvB.sample_posterior_over_accuracy(anno_iv.annotations, 200, burn_in_samples=100, thin_samples=3)
ci_ctv_mean = samples_ctvB[0].mean(axis=0)
print(ci_ctv_mean)

Now we will package up the predicted data into a format we can use for other, subsequent analysis:

In [ ]:
print(posterior_dvB.shape)
print(posterior_ivB.shape)
print(posterior_cvB.shape)
print(posterior_ctvB.shape)

In [ ]:
predicted_annotations = np.concatenate((posterior_dvB, posterior_ivB, posterior_cvB, posterior_ctvB), axis=1) # posterior_dvBt, posterior_ivBt, posterior_cvBt, posterior_ctvBt), axis=1)

In [ ]:
predicted_annotations.shape

These annotations allowed us to uncover the degree to which social scientists alter their models to achieve a better fit...undocumented data mining. The answer was that social scientists did mine their data, but that it likely improved their analysis because change in the social world was the result of greater distortion than undocumented data mining.

## Another example analysis looks at a different data set of Hotel Reviews by a variety of patrons.

In [ ]:
df_hotels = pandas.read_csv('/content/hot_Reviews.csv', index_col=0)
df_hotels[:5]

Here a rank of 0 is a missing value and to simplify things more we will convert from a 1-10 scale to a 1-5 scale, with 0 as missing

In [ ]:
df_hotels = df_hotels.apply(lambda x: x // 2) #integer divide by 2 rounds all values

And we can visualize all the reviews as a heatmap with the missing values greyed out

In [ ]:
fig, ax = plt.subplots(figsize = (20,20))
seaborn.heatmap(df_hotels, cmap='rainbow', ax = ax)
plt.show()

To give the dataframe to pyanno we need to convert to np array and change the nans to intergers, lets use -1

In [ ]:
hot_mat = np.array(df_hotels.fillna(-1).as_matrix())
anno_hot = AnnotationsContainer.from_array(hot_mat, missing_values=[-1])

In [ ]:
anno_hot.annotations

In [ ]:
anno_hot.labels

In [ ]:
anno_hot.missing_values

Look at coder agreement

In [ ]:
pyanno.measures.agreement.labels_frequency(anno_hot.annotations, 6)#6 possible catagories

In [ ]:
c = pyanno.measures.agreement.confusion_matrix(anno_hot.annotations[:,0], anno_hot.annotations[:,1], 6) #6 possible catagories
print(c)
ac = seaborn.heatmap(c)
plt.show()

Most agreement is on 2 i.e. an average hotel and there's little agreement as rating go higher, likely due to scarcity in the sample

In [ ]:
scotts_pi(anno_hot.annotations[:,0], anno_hot.annotations[:,1])

In [ ]:
krippendorffs_alpha(anno_hot.annotations[::])

In [ ]:
cohens_kappa(anno_hot.annotations[:,0], anno_hot.annotations[:,1])

In [ ]:
m = pairwise_matrix(cohens_kappa, anno_hot.annotations)
fig, ax = plt.subplots(figsize = (15, 15))
seaborn.heatmap(m, ax =ax)

In [ ]:
model_hot = ModelBt.create_initial_state(6, 49)
model_hot.mle(anno_hot.annotations)
print(model.theta)
print(model_hot.log_likelihood(anno_hot.annotations))

In [ ]:
def makeQuestionComparison(model, anno_target, num_questions = 20):
    votes = []
    for r in anno_target.annotations:
        v = [0] * len(anno_target.labels)
        votes.append(v)
        for a in r:
            if a > -1:
                v[a] += 1
    votes_array = np.array(votes)
    posterior = model.infer_labels(anno_target.annotations)
    fig, (ax1, ax2) = plt.subplots(ncols=2, figsize = (15, 10), sharey=True)

    seaborn.heatmap(votes_array[:num_questions], annot = True, ax=ax2)
    seaborn.heatmap(np.nan_to_num(posterior,0)[:num_questions], annot=True, ax =ax1)
    ax1.set_title("Model")
    ax2.set_title("Votes")
    return fig, (ax1, ax2)

In [ ]:
makeQuestionComparison(model_hot, anno_hot)

# Homework

In this notebook we explored a variety of sampling methods, as well as its applications to machine learning and deep learning. We then explored annotations of your data (these are what power our models!), and how to identify reliable annotations. These concepts will be crucial as we discover active learning in the next tutorial. 

In this notebook's exercises, you will be using these concepts in the context of your social scientific data and experiments.

**1)** Run 3 probabilistic sampling methods and 2 non-probabilistic methods and explore the samples returned. How would sampling help with your data?

In [ ]:
sampling_help = 'value' #@param {type:"string"}

**2)** Find an imbalanced dataset and build a classifier to predict the label causing the imbalance. Explore undersampling and oversampling solutions to your dataset.

**3)** Use bootstrap sampling to test the stability of a word, sentence, or graph embedding.

**4)** Perform a content annotation survey of some kind in which at least 3 people evaluate and code each piece of content, using Amazon Mechanical Turk as described in the link below, or by hand with friends.  With the resulting data, calculate, visualize and discuss inter-coder agreement or covariation with appropriate metrics. What does this means for the reliability of human assessments regarding content in your domain?

Run your own survey and annotations: [MTurk tutorial](https://canvas.uchicago.edu/courses/33672/files/4767031/download).



In [ ]:
reliability = 'value' #@param {type:"string"}


**5)** In the cells immediately following, use the results of your content annotation survey to predict high and low-quality analysts, then predict MAP estimates for your codes in question. What do these estimates suggest about the distribution of skill among your coders? How different are these estimates from a majority vote?

In [ ]:
reliability = 'value' #@param {type:"string"}